src/mc-fle2-payload-ieev.c

```diff
@@ -0,0 +1,341 @@
+                  "called after mc_FLE2IndexedEqualityEncryptedValue_parse");
+      return NULL;
+   }
+
+   return &ieev->S_KeyId;
+}
+
+bool
+mc_FLE2IndexedEqualityEncryptedValue_add_S_Key (
+   _mongocrypt_crypto_t *crypto,
+   mc_FLE2IndexedEqualityEncryptedValue_t *ieev,
+   const _mongocrypt_buffer_t *S_Key,
+   mongocrypt_status_t *status)
+{
+   if (!ieev->parsed) {
+      CLIENT_ERR ("mc_FLE2IndexedEqualityEncryptedValue_add_S_Key must be "
+                  "called after mc_FLE2IndexedEqualityEncryptedValue_parse");
+      return false;
+   }
+
+   if (ieev->inner_decrypted) {
+      CLIENT_ERR ("mc_FLE2IndexedEqualityEncryptedValue_add_S_Key must not be "
+                  "called twice");
+      return false;
+   }
+
+   /* Attempt to decrypt InnerEncrypted */
+   if (S_Key->len != MONGOCRYPT_KEY_LEN) {
+      CLIENT_ERR ("mc_FLE2IndexedEqualityEncryptedValue_add_S_Key expected "
+                  "S_Key to be %d bytes, got: %" PRIu32,
+                  MONGOCRYPT_KEY_LEN,
+                  S_Key->len);
+      return false;
+   }
+
+   /* Get the TokenKey from the last 32 bytes of S_Key */
+   _mongocrypt_buffer_t TokenKey;
+   if (!_mongocrypt_buffer_from_subrange (&TokenKey,
+                                          S_Key,
+                                          S_Key->len - MONGOCRYPT_TOKEN_KEY_LEN,
+                                          MONGOCRYPT_TOKEN_KEY_LEN)) {
+      CLIENT_ERR ("mc_FLE2IndexedEqualityEncryptedValue_add_S_Key unable to "
+                  "parse TokenKey from S_Key");
+      return false;
+   }
+
+   /* Use TokenKey to create ServerDataEncryptionLevel1Token and decrypt
+    * InnerEncrypted. */
+   {
+      mc_ServerDataEncryptionLevel1Token_t *token =
+         mc_ServerDataEncryptionLevel1Token_new (crypto, &TokenKey, status);
+      if (!token) {
+         return false;
+      }
+
+      const _mongocrypt_buffer_t *token_buf =
+         mc_ServerDataEncryptionLevel1Token_get (token);
+      uint32_t bytes_written;
+
+      _mongocrypt_buffer_resize (
+         &ieev->Inner,
+         _mongocrypt_fle2_calculate_plaintext_len (ieev->InnerEncrypted.len));
+
+      /* Decrypt InnerEncrypted. */
+      if (!_mongocrypt_fle2_do_decryption (crypto,
+                                           token_buf,
+                                           &ieev->InnerEncrypted,
+                                           &ieev->Inner,
+                                           &bytes_written,
+                                           status)) {
+         mc_ServerDataEncryptionLevel1Token_destroy (token);
+         return false;
+      }
+      mc_ServerDataEncryptionLevel1Token_destroy (token);
+   }
+
+   /* Parse Inner for K_KeyId. */
+   uint32_t offset = 0;
+   /* Read uint64_t length. */
+   if (offset + 8 > ieev->Inner.len) {
+      CLIENT_ERR ("mc_FLE2IndexedEqualityEncryptedValue_add_S_Key expected "
+                  "Inner byte length: %" PRIu32 " got: %" PRIu32,
```

Comment:
Comparison is greater-than, but error message suggests equality expected. Should error message be changed to indicate greater-than expectation? Same question applies to similar greater-than comparisons followed by `CLIENT_ERR` messages in this file.

Reply:
Updated error messages to note the expectation is `>=`.